<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Mendelbase.jl : MendelBase contains the base functions of OpenMendel.">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Mendelbase.jl</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/OpenMendel/MendelBase.jl">View on GitHub</a>

          <h1 id="project_title">Mendelbase.jl</h1>
          <h2 id="project_tagline">MendelBase contains the base functions of OpenMendel.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/OpenMendel/MendelBase.jl/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/OpenMendel/MendelBase.jl/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="overview" class="anchor" href="#overview" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overview</h3>

<p><a href="https://github.com/OpenMendel/MendelBase.jl">MendelBase</a> is a utility package that contains the base functions of <a href="https://openmendel.github.io">OpenMendel</a>. It provides the information needed to process keywords and read data from external files. MendelBase is necessary to run any analysis package, and you should install MendelBase first, before you install the other packages.</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation</h3>

<p>Within Julia, use the package manager to install MendelBase:</p>

<pre><code>Pkg.clone("git@github.com:OpenMendel/MendelBase.jl.git")
</code></pre>

<p>This package supports Julia v0.4.</p>

<h3>
<a id="input-files" class="anchor" href="#input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Input Files</h3>

<p>To make your genetic data available to OpenMendel, the various input files must be in a format
that OpenMendel can read. To provide OpenMendel with the information it needs to perform its analysis you must prepare the following files. Examples input files can be found in the docs subfolders of the different analysis packages. (An analysis option won't always need every file type.)</p>

<ul>
<li>
<a href="#control-file">Control File</a>: Specifies the names of your data input and output files and any optional parameters (<em>keywords</em>) for the analysis. (For a complete list of keywords, see <a href="#keywords-table">Keywords Table</a> below).</li>
<li>
<a href="#locus-file">Locus File</a>: Names and describes the genetic loci in your data.</li>
<li>
<a href="#pedigree-file">Pedigree File</a>: Gives information about your individuals, such as name, sex, family structure, and ancestry.</li>
<li>
<a href="#phenotype-file">Phenotype File</a>: Lists the available phenotypes.</li>
<li>
<a href="#snp-definition-file">SNP Definition File</a>: Defines your SNPs with information such as SNP name, chromosome, position, allele names, allele frequencies. This may be a PLINK .bim format file.</li>
<li>
<a href="#snp-data-file">SNP Data File</a>: Holds the genotypes for your data set. Must be a standard binary PLINK .bed file in SNP major format. If you have a SNP data file you must have a SNP definition file.</li>
</ul>

<h3>
<a id="control-files" class="anchor" href="#control-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Control files<a id="control-file"></a>
</h3>

<p>The format of the Control file is</p>

<pre><code>Keyword = Keyword_Value(s)
</code></pre>

<p>Below is an example Control file to run GWAS (text after "#" is ignored and treated as comments):</p>

<pre><code>#
# Input and Output files.
#
plink_input_basename = gwas 1 data
output_file = gwas 1 Output.txt
manhattan_plot_file = ManPlotOutput_1.png
#
# Analysis parameters for GWAS option.
#
regression = linear
regression_formula = Trait ~ Sex
</code></pre>

<p>In the example above, there are five keywords. The first three specify input and output files: <em>plink_input_basename</em>, <em>output_file</em>, <em>manhattan_plot_file</em>. The last two specify analysis parameters: <em>regression</em> and <em>regression_formula</em>. The text after the '=' are the keyword values. For keywords that can take a set of values, list the values separated by commas.</p>

<h3>
<a id="keywords" class="anchor" href="#keywords" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Keywords<a id="keywords-table"></a>
</h3>

<p>This is a list of the general OpenMendel keywords common to most analysis package. The keywords specific to an analysis package are covered in detail in the documentation for each package. The names of keywords are <em>not</em> case sensitive. (The keyword <em>values</em> may be case sensitive.)</p>

<table>
<thead>
<tr>
<th>Keyword</th>
<th>Default Value</th>
<th>Allowed Values</th>
<th>Short Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>affected_designator</td>
<td>affected</td>
<td></td>
<td>Disease phenotype designator</td>
</tr>
<tr>
<td>allele_pseudo_count</td>
<td>0.1</td>
<td>Real</td>
<td>Number of prior alleles or haplotypes</td>
</tr>
<tr>
<td>allele_separator</td>
<td>/</td>
<td>1 Character</td>
<td>Separates codominant alleles</td>
</tr>
<tr>
<td>complexity_threshold</td>
<td>5.00E+07</td>
<td>Real</td>
<td>Criterion for not analyzing pedigrees</td>
</tr>
<tr>
<td>disease_status</td>
<td></td>
<td></td>
<td>Name of column designating disease status</td>
</tr>
<tr>
<td>distribution</td>
<td>Normal()</td>
<td>Binomial(), Gamma(), Normal(), or Poisson()</td>
<td>Name of distribution</td>
</tr>
<tr>
<td>eliminate_genotypes</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Whether or not genotypes are eliminated</td>
</tr>
<tr>
<td>female</td>
<td>"female", "f", 'f', "2", '2', 2</td>
<td>Female designators separated by commas</td>
<td>Female designators</td>
</tr>
<tr>
<td>field_separator</td>
<td>','</td>
<td>1 Character</td>
<td>Character used to separates values in data files</td>
</tr>
<tr>
<td>genetic_map_function</td>
<td>Haldane</td>
<td>Haldane or Kosambi</td>
<td>Genetic map units</td>
</tr>
<tr>
<td>link</td>
<td>IdentityLink()</td>
<td></td>
<td>Name of link function</td>
</tr>
<tr>
<td>locus_file</td>
<td></td>
<td></td>
<td>Name of input locus file</td>
</tr>
<tr>
<td>lump_alleles</td>
<td>FALSE</td>
<td>TRUE, FALSE</td>
<td>Whether or not alleles are combined</td>
</tr>
<tr>
<td>male</td>
<td>"male", "m", 'm', "1", '1', 1</td>
<td>Male designators separated by commas</td>
<td>Male designators</td>
</tr>
<tr>
<td>new_pedigree_file</td>
<td></td>
<td></td>
<td>OpenMendel generated output pedigree file</td>
</tr>
<tr>
<td>ordered_allele_separator</td>
<td>| (vertical bar)</td>
<td>1 Character</td>
<td>Separates ordered codominant alleles</td>
</tr>
<tr>
<td>output_file</td>
<td>Mendel_Output.txt</td>
<td></td>
<td>Name of standard results output file</td>
</tr>
<tr>
<td>pedigree_file</td>
<td></td>
<td></td>
<td>Name of input pedigree file</td>
</tr>
<tr>
<td>phenotype_file</td>
<td></td>
<td></td>
<td>Name of input phenotype file</td>
</tr>
<tr>
<td>plink_input_basename</td>
<td></td>
<td></td>
<td>Basename of PLINK input files</td>
</tr>
<tr>
<td>plink_output_basename</td>
<td></td>
<td></td>
<td>Basename of PLINK output files</td>
</tr>
<tr>
<td>populations</td>
<td></td>
<td>Names of populations separated by commas</td>
<td>Names of populations</td>
</tr>
<tr>
<td>product_mode</td>
<td>TRUE</td>
<td>TRUE, FALSE</td>
<td></td>
</tr>
<tr>
<td>seed</td>
<td>1234</td>
<td>Integer</td>
<td>Random number generator seed</td>
</tr>
<tr>
<td>snpdata_file</td>
<td></td>
<td></td>
<td>Name of input SNP data file</td>
</tr>
<tr>
<td>snpdefinition_file</td>
<td></td>
<td></td>
<td>Name of input SNP definition_file</td>
</tr>
</tbody>
</table>

<h3>
<a id="general-data-file-format" class="anchor" href="#general-data-file-format" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>General Data File Format</h3>

<p>All data files are in text file format with one exception: if you have dense SNP data, the SNP data should be in binary files in PLINK <strong>.bed</strong> file format. OpenMENDEL will also accept <a href="http://pngu.mgh.harvard.edu/%7Epurcell/plink/data.shtml">PLINK format</a> <strong>.fam</strong> and <strong>.bim</strong>. All non-PLINK OpenMendel data files list data in columns, with a header row with field names for each column. The column need not be in any fixed order. Many column headers have required, case sensitive names. The column separators can be any single character, specified in the control file with the <strong>field_separator</strong>. The the default separator is <strong>comma</strong>. If a data element contains the field separator (e.g. comma) it must be enclosed in double quotes, or it will be read as a new field. Missing data is blank ("") or "NA". The PLINK file missing data symbols are "0" or "-9".</p>

<h3>
<a id="locus-file" class="anchor" href="#locus-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Locus File<a id="locus-file"></a>
</h3>

<p>The Locus file defines your genetic markers unless you are working with a dense SNP map, in which case you must use a SNP Definition. The Locus file must have a header row with the field names (in any order): Locus, Allele, Chromosome, Frequency. If your samples come from different populations, instead of one column labeled Frequency, you may name the populations in the control file with the keyword <em>populations</em>, and give an allele frequency for each population. The population frequency field names are the population names specified in the control file. The population frequencies need not sum to 1, but will be normalized to 1 in the analysis. You may also include locus position information with optional fields <strong>Morgans</strong>, <strong>FemaleMorgans</strong> and <strong>MaleMorgans</strong>, and <strong>Basepairs</strong>. For each locus there will be one line per locus allele.</p>

<p><strong>Example</strong></p>

<pre><code>Locus,Allele,Chromosome,European,African,Chinese
ABO,A,9,0.27,0.18,0.19
ABO,B,9,0.06,0.11,0.17
ABO,O,9,NA,0.71,0.64
Rh,D,1,0.61,0.81,1
Rh,d,1,0.39,0.19,0
SNP,1,9,NA,NA,NA
SNP,2,9,NA,NA,NA
XSNP,1,X,0.3,0.4,0.5
XSNP,2,X,0.7,0.6,0.5
</code></pre>

<h3>
<a id="pedigree-file" class="anchor" href="#pedigree-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pedigree File<a id="pedigree-file"></a>
</h3>

<p>A Pedigree file is always required. It can be in PLINK FAM file format, in which case it must have the extension .fam, or it can have the format of a required header rows with the following fieldnames (in any order): Pedigree,Person,Mother,Father,Sex,Twin, and optional columns such as population ancestries. The fractional ancestries need not sum to 1, but will be normalized to 1 in the analysis.</p>

<p><strong>Example</strong></p>

<pre><code>Pedigree,Person,Mother,Father,Sex,Twin,BirthYear,Politician,European,African,Chinese,ABO,Rh,Xg,XSNP,HtMeters
Clinton,Bill,,,male,0,1946.0,true,1.0,0.0,0.0,A,+,+,2/2,1.88
Clinton,Hillary,,,female,0,1947.0,true,1.0,0.0,0.0,O,-,-,1/1,NA
Clinton,Chelsea,Hillary,Bill,female,0,1980.0,false,NA,NA,NA,A,+,-,1/2,NA
Obama,Barack,,,male,0,1961.0,true,0.5,0.5,0.0,O,+,NA,1/1,1.85
Obama,Michelle,,,female,0,1964.0,false,0.3,0.7,0.0,B,+,NA,2/2,NA
Obama,Malia,Michelle,Barack,female,0,1998.0,false,NA,NA,NA,O,-,+,1/2,NA
Obama,Sasha,Michelle,Barack,female,0,2001.0,false,NA,NA,NA,B,+,+,1/2,NA
</code></pre>

<h3>
<a id="phenotype-file" class="anchor" href="#phenotype-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Phenotype File<a id="phenotype-file"></a>
</h3>

<p>Format</p>

<p><strong>Example</strong></p>

<pre><code>Locus,Phenotype,Genotypes
ABO,A,"A/A,A/O"
ABO,B,"B/B,B/O"
ABO,AB,"A/B"
ABO,O,"O/O"
Rh,+,"D/D,D/d"
Rh,-,"d/d"
Xg,+,"+/+,+/-"
Xg,-,"-/-"
</code></pre>

<h3>
<a id="snp-definition-file" class="anchor" href="#snp-definition-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP Definition File<a id="snp-definition-file"></a>
</h3>

<p>Defines your SNPs with information such as SNP name, chromosome, position, allele names, allele frequencies. This may be a PLINK .bim format file.</p>

<h3>
<a id="snp-data-file" class="anchor" href="#snp-data-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>SNP Data File<a id="snp-data-file"></a>
</h3>

<p>Holds the genotypes for your data set. Must be a standard binary PLINK .bed file in SNP major format. If you have a SNP data file you must have a SNP definition file.</p>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citation</h3>

<p>If you use this analysis package in your research, please cite the following reference in the resulting publications:</p>

<p><em>Lange K, Papp JC, Sinsheimer JS, Sripracha R, Zhou H, Sobel EM (2013) Mendel: The Swiss army knife of genetic analysis programs. Bioinformatics 29:1568-1570.</em></p>



<h3>
<a id="acknowledgments" class="anchor" href="#acknowledgments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>This project is supported by the National Institutes of Health under NIGMS awards R01GM053275 and R25GM103774 and NHGRI award R01HG006139.</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Mendelbase.jl maintained by <a href="https://github.com/OpenMendel">OpenMendel</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
